src/SFML/Window/Android/JoystickImpl.cpp

```diff
@@ -83,10 +179,74 @@ Joystick::Identification JoystickImpl::getIdentification() const
 
 
 ////////////////////////////////////////////////////////////
-JoystickState JoystickImpl::update()
+JoystickState JoystickImpl::update() const
+{
+    // Retrieve activity states
+    ActivityStates&       states = getActivity();
+    const std::lock_guard lock(states.mutex);
+
+    JNIEnv* env = states.activity->env;
+    auto    jni = Jni::attachCurrentThread(states.activity->vm, &env);
+    if (!jni)
+    {
+        err() << "Failed to initialize JNI" << std::endl;
+        return {false};
+    }
+
+    auto inputDeviceClass = JniInputDeviceClass::findClass(env);
+    if (!inputDeviceClass)
+    {
+        return {false};
+    }
+
+    const bool isConnected = inputDeviceClass->getDevice(m_currentDeviceIdx).has_value();
+    if (states.joystickStates.find(m_currentDeviceIdx) == states.joystickStates.end())
+    {
+        // This technically shouldn't happen, but when I have connect physical gamepad
+        // and then connect/disconnect a bluetooth one, states for the physical one
+        // disappears for a frame and then it reconnects.
+        return {false};
+    }
+
+    return {isConnected,
+            states.joystickStates.at(m_currentDeviceIdx).axes,
+            states.joystickStates.at(m_currentDeviceIdx).buttons};
```

Comment:
Direct access should be fine, as we just check it above

```suggestion
            states.joystickStates[m_currentDeviceIdx].axes,
            states.joystickStates[m_currentDeviceIdx].buttons};
```